rfcs/react-components/convergence/custom-styling.md

```diff
@@ -0,0 +1,214 @@
+  {},
+);
+
+export const ComponentCustomStylesContextProvider = ComponentCustomStylesContext.Provider;
+```
+
+When a caller wanted to customize the component, they use the provider. Callers would be able to control custom
+styling scope just like Fluent Provider. While they would need to use a provider per custom styled component,
+these could be aggregated into a component that transcludes children.
+
+```tsx
+<App>
+  <FluentProvider theme={webLightTheme}>
+    <ComponentCustomStylesContextProvider value={{ useCustomStyles }}>
+      {/* other component hierarchy */}
+      <Component>Hello custom styles!</Component>
+    </ComponentCustomStylesContextProvider>
+  </FluentProvider>
+</App>
+```
+
+### Open issue: Centralized and singular custom styling method
+
+It would be nice to have a single method on context that could apply for all components. This proves difficult
+with each component having its own props and state types.
+
+Another approach would be to define single context value type in the react-shared-contexts package that had
+a method for each component, but that causes a circular dependency problem for components.
+
+There may some typescript trickery that would allow for a single context value type, but having each component define
+the context type it will inspect for custom styling follows SOLID dependency inversion.
+
+### Pros
+
+- Custom styling gets the same power as the default component useStyles hook.
+- This builds on using slots as the addressable items for applying styles.
+- The custom styles can be built using Griffel with build-time optimization.
+- Can be applied at different scope like FluentProvider.
+- Does not require composing or introduce a new component type that must be adopted in lieu of the default component.
+- Fewer CSS specificity problems as styles can get overwritten.
+- Provides the type-safety expected from Fluent UI React.
+- Minimal code to add to each component. Could be put into the code generation template.
+- Does not significantly effect rendering performance.
+- Partners building components can follow the same practice to support custom styling in their libraries.
+
+### Cons
+
+- Custom styling all components requires adding and configuration lots of context providers.
+- Requires writing new code for every component, adding a story, and documenting the feature.
+- Context is harder to discover as it is outside the props contract.
+- Callers doing custom styling may still take a dependency on the CSS internals of the component if they don't replace a
+  style block completely.
+
+## Option B: Demonstrate recomposition for partners to define their own style customization systems
+
+The hooks composition model was architected to separate concerns of component behavior, style, and rendering.
+Rather than provide an additional mechanism for custom styling, this option would assist partners with leveraging
```

Comment:
I don't think token system scales well enough to meet the needs of apps with hundreds of engineers and designers.

- It assumes tokens are always applied in a semantically correct way. I know this is not a safe assumption. I see designers and engineers mis-applying semantic slots in v8 all the time. Not ideal, but it's reality. For tokens to scale effectively, Designers and Engineers will need to know with 100% certainty across several libraries which tokens are being used to prevent regressions. The issue gets even more complicated when end feature developers start consuming themes from multiple Fluent extensions.

- Teams are going to run out of headroom real quick if they're only able to customize via 300 tokens. For color alone, MADS uses about 150 semantic slots because the base 70 we get from Fluent weren't enough. Add more for corner radius, elevation, thicknesses, etc and we'll run out of accurately named tokens really quick.

Providing a centralized way to override the styles of a specific component provides a complimentary solution to token  based styling.

Reply:
FWIW, the hook recomposition approach is different than adding new tokens. What you wrote does apply to adding more tokens to the system and the theme fragility problem is tough to solve.  When a new component recomposes another one using hooks, the style hook should use the same set of design tokens available in v9.

Reply:
Rereading - good point.  Resolving this conversation as the content has been updated to reflect the concern over 300 tokens. If someone goes over 300 tokens, the system still works it just performs a little more slowly and bundling size slows initial load of the application.